src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp

```diff
@@ -304,11 +304,109 @@ class StaticSlice {
   }
 };
 
-struct NoopSlice final : OpRewritePattern<mlir::stablehlo::SliceOp> {
-  using OpRewritePattern::OpRewritePattern;
+LogicalResult failIfDynamicShape(Operation *op, PatternRewriter &rewriter) {
+  for (auto result : op->getResults()) {
+    auto type = dyn_cast<RankedTensorType>(result.getType());
+    if (!type || !type.hasStaticShape())
+      return rewriter.notifyMatchFailure(
+          op, "unsupported dynamic shape for output.");
+  }
+
+  for (auto operand : op->getOperands()) {
+    auto type = dyn_cast<RankedTensorType>(operand.getType());
+    if (!type || !type.hasStaticShape())
+      return rewriter.notifyMatchFailure(
+          op, "unsupported dynamic shape for input.");
+  }
+
+  return success();
+}
 
-  LogicalResult matchAndRewrite(mlir::stablehlo::SliceOp op,
+LogicalResult failIfFuncOpInterfaceHasAttr(Operation *op,
+                                           std::optional<StringRef> attrName,
+                                           PatternRewriter &rewriter) {
+  if (!attrName)
+    return success();
+
+  if (auto func = op->template getParentOfType<FunctionOpInterface>()) {
+    if (func->template hasAttrOfType<UnitAttr>(*attrName))
+      return rewriter.notifyMatchFailure(op, "disabled by attribute.");
+  }
+
+  return success();
+}
+
+template <typename OpTy>
+struct CheckedOpRewritePattern : public OpRewritePattern<OpTy> {
+  using Base = OpRewritePattern<OpTy>;
+  using Base::Base;
+
+  LogicalResult matchAndRewrite(OpTy op,
                                 PatternRewriter &rewriter) const override {
+    LogicalResult res =
+        failIfFuncOpInterfaceHasAttr(op, disablePatternAttrName(), rewriter);
+    if (res.failed())
+      return res;
+
+    if (!supportsDynamicShapes()) {
```

Comment:
change this to OpTy::supportsDynamicShapes, then change `supportsDynamicShapes` to be a static member function [not a virtual one], and overrides also to be new static member function. This means that we don't require virtual function indirection [and this becomes a constant check]

Reply:
can you check if the latest changes are what you wanted?

Reply:
yeah indeed!